docs/RingBuffer.md

```diff
@@ -0,0 +1,414 @@
+# New ebpf Ring Buffer Map (proposal)
+
+## Overview
+
+The current ringbuffer uses a pure callback-based approach to reading the ringbuffer.
+Linux also supports memory-mapped polling consumers, which can't be directly supported in the current model.
+
+The new API will support 2 consumer types: callbacks and direct access to the mapped producer memory (with poll to wait for data).
+
+Callback consumer:
+
+1. Call `ring_buffer__new` to set up callback.
+2. The callback will be invoked for each record written to the ring buffer.
+
+Mapped memory consumer:
+
+1. Call `ebpf_ring_buffer_get_buffer` to get pointers to the mapped producer/consumer pages.
+2. Call `ebpf_ring_buffer_get_wait_handle` to get the wait handle.
+3. Directly read records from the producer pages (and update consumer offset as we read).
+4. Call `WaitForSingleObject`/`WaitForMultipleObject` as needed to wait for new data to be available.
+
+### Differences from linux API
+
+Linux has only polling ring buffer consumers, even when callbacks are used.
+On linux the user code can call `ring_buffer__consume()` to invoke the callback on all available data,
+and `ring_buffer__poll()` to wait for data if needed and then consume available data.
+Linux consumers can also directly read from the mapped memory by using `mmap()` to map the data
+into user space and `ring_buffer__epoll_fd()` to get an epoll wait handle.
+
+On Windows asynchronous events are supported by default,
+so nothing extra needs to be done for callbacks to be invoked.
+
+If the `RINGBUF_FLAG_NO_AUTO_CALLBACK` flag is set, callbacks will not automatically be called and `ring_buffer__poll()`
```

Comment:
Why should the default be different between Windows vs Linux?
Why not a `RINGBUF_FLAG_AUTO_CALLBACK` flag instead?